### ISSUE TITLE
Preserving order with serde

### ISSUE THREAD

**Issue by cisaacson** [bluss/indexmap#235]
We have a nested map in a JSON result, and a struct to handle that, like this:

```
#[derive(Serialize, Deserialize, Debug)]
struct MyStruct {
    name: String,
    avg_value: f64,
    value_map: IndexMap<String, f64>,
}
```

I need to be sure that the value_map insertion order will be preserved. In my initial test it does preserve insertion order. It would seem this is the whole purpose of `IndexMap` but I was reading about `#[serde(with = "indexmap::serde_seq")]` to make the guarantee, and when I tried that it did not work. I got this error:

`invalid type: map, expected a sequenced map`

Without that it all runs fine, and as I said the initial test does preserve the order. Please confirm this is the correct way to do this.

**Comment by cuviper** [comment]
For its part, `indexmap` does emit items in order while serializing, and it also preserves the order it sees while deserializing just like a sequence of `insert` calls. The reason `serde_seq` exists is that some serialization formats don't preserve the order, like `flexbuffers` per #156, so we give the option to output a strict sequence instead of a serde map.

If your format already preserves order, I don't think you need to bother with `serde_seq` here. For example, `serde_json` has an explicit `"preserve_order"` feature for that -- which even uses `IndexMap` to achieve it in their `Value` type. :slightly_smiling_face:  

> `invalid type: map, expected a sequenced map`

I think that means you're trying to use `serde_seq` to deserialize something that was serialized as a normal serde map, which includes JSON objects.

I wonder if we could just let `serde_seq` be flexible in deserializing things either way, but I fear that could disrupt use cases that depend on strictness. IIRC, serde's support for untagged enums will try deserializing one variant and then another. If we're more flexible, we could end up eating data that would have fallen through to another variant otherwise.

**Comment by cisaacson** [comment]
Thanks @cuviper very much for the quick feedback. All we need in our case is to keep the `IndexMap` in the same order as the original JSON, it appears to be doing that now.

I tried to use the `json_serde::preserve_order` feature, and that only supports `Map<String, Value>` so it is not nearly as convenient as what `IndexMap` on its own offers. We want to serialize into our own struct, rather than a JSON `Value`.

So in summary it seems like it is working as expected, unless you have any other concerns I'll close the ticket.

**Comment by cuviper** [comment]
Yes, `IndexMap` will respect the order it is given, so you just need to make sure your JSON layer does too.
